File: project_euler_unified/problem_035/solution.py
"""Project Euler Problem 35: Circular Primes"""

def sieve(limit):
    is_prime = [True] * limit
    is_prime[0] = is_prime[1] = False
    for i in range(2, int(limit**0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, limit, i):
                is_prime[j] = False
    return is_prime

def main():
    N = 1_000_000
    is_prime = sieve(N)
    count = 0
    for p in range(2, N):
        if not is_prime[p]:
            continue
        s = str(p)
        k = len(s)
        if all(is_prime[int(s[i:] + s[:i])] for i in range(k)):
            count += 1
    print(count)  # 55

if __name__ == "__main__":
    main()

File: project_euler_unified/problem_035/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    const int N = 1000000;
    vector<bool> is_prime(N, true);
    is_prime[0] = is_prime[1] = false;
    for (int i = 2; i * i < N; i++)
        if (is_prime[i])
            for (int j = i * i; j < N; j += i)
                is_prime[j] = false;

    int count = 0;
    for (int p = 2; p < N; p++) {
        if (!is_prime[p]) continue;
        int k = 0, tmp = p;
        while (tmp > 0) { k++; tmp /= 10; }
        int pw = 1;
        for (int i = 0; i < k - 1; i++) pw *= 10;
        bool circular = true;
        int r = p;
        for (int i = 0; i < k; i++) {
            if (r < 0 || r >= N || !is_prime[r]) { circular = false; break; }
            r = (r % 10) * pw + r / 10;
        }
        if (circular) count++;
    }
    cout << count << endl;  // 55
    return 0;
}
